File: server/session.py
"""Session management: start, save, end, and record events."""
import os
import json
import time
from datetime import datetime, timezone
from . import state
import logging
import asyncio
from .ai_client import analyze_prompt

logger = logging.getLogger(__name__)


def start_session(user='juedwards'):
    # Use the shared state module to set session attributes so all modules see updates
    state.session_start_time = state.now_utc()
    state.session_id = f"minecraft_session_{state.session_start_time.strftime('%Y%m%d_%H%M%S')}"
    state.session_events = []
    initial_event = {
        'timestamp': state.session_start_time.isoformat(),
        'event_type': 'session_start',
        'data': {'session_id': state.session_id, 'server_version': '1.0', 'user': user}
    }
    state.session_events.append(initial_event)
    state.session_file = os.path.join(state.DATA_DIR, f"{state.session_id}.json")
    save_session_realtime()
    logger.info(f"📝 Started new session: {state.session_id}")


def save_session_realtime():
    if not state.session_file:
        return
    current_time = state.now_utc()
    session_data = {
        'session_info': {
            'id': state.session_id,
            'start_time': state.session_start_time.isoformat() if state.session_start_time else None,
            'last_update': current_time.isoformat(),
            'duration_seconds': (current_time - state.session_start_time).total_seconds() if state.session_start_time else 0,
            'total_events': len(state.session_events),
            'user': 'juedwards',
            'status': 'active' if state.active_players else 'ended'
        },
        'events': state.session_events
    }
    temp_file = f"{state.session_file}.tmp"
    with open(temp_file, 'w', encoding='utf-8') as f:
        json.dump(session_data, f, indent=2)
    if os.path.exists(state.session_file):
        os.remove(state.session_file)
    os.rename(temp_file, state.session_file)
    state.last_save_time = time.time()


def end_session():
    if not state.session_id:
        return
    session_end_time = state.now_utc()
    state.session_events.append({
        'timestamp': session_end_time.isoformat(),
        'event_type': 'session_end',
        'data': {
            'session_id': state.session_id,
            'duration_seconds': (session_end_time - state.session_start_time).total_seconds() if state.session_start_time else 0,
            'total_events': len(state.session_events)
        }
    })
    save_session_realtime()
    logger.info(f"💾 Session ended and saved: {state.session_file} ({len(state.session_events)} events)")


def record_event(event_type, data):
    if state.session_id:
        event = {'timestamp': state.now_utc().isoformat(), 'event_type': event_type, 'data': data}
        state.session_events.append(event)
        state.event_buffer.append(event)
        current_time = time.time()
        if (current_time - (state.last_save_time or 0) > 5) or (len(state.event_buffer) >= 50):
            save_session_realtime()
            state.event_buffer = []
            state.last_save_time = current_time
            logger.debug(f"💾 Auto-saved session with {len(state.session_events)} events")


async def analyze_player_data():
    """Analyze current player data against rubric using the ai_client.analyze_prompt wrapper."""
    global latest_assessment_results
    try:
        rubric_path = os.path.join(os.path.dirname(__file__), '..', 'rubric.md')
        if not os.path.exists(rubric_path):
            return {'error': 'Rubric file not found'}

        with open(rubric_path, 'r', encoding='utf-8') as f:
            rubric_content = f.read()

        # Aggregate events by player
        player_analysis_data = {}
        for event in state.session_events:
            if event['event_type'] in ['player_position', 'block_placed', 'block_broken', 'player_join', 'player_leave', 'player_chat']:
                pdata = event.get('data', {})
                player_id = pdata.get('player_id') or pdata.get('player_name')
                player_name = pdata.get('player_name', player_id)
                if player_name not in player_analysis_data:
                    player_analysis_data[player_name] = {'positions': [], 'blocks_placed': [], 'blocks_broken': [], 'join_time': None, 'leave_time': None, 'total_distance': 0}

                entry = player_analysis_data[player_name]
                etype = event['event_type']
                if etype == 'player_position':
                    pos = pdata.get('position')
                    if pos:
                        entry['positions'].append(pos)
                elif etype == 'block_placed':
                    entry['blocks_placed'].append({'type': pdata.get('block_type'), 'position': pdata.get('estimated_block_position'), 'time': event.get('timestamp')})
                elif etype == 'block_broken':
                    entry['blocks_broken'].append({'type': pdata.get('block_type'), 'position': pdata.get('estimated_block_position'), 'time': event.get('timestamp')})
                elif etype == 'player_join':
                    entry['join_time'] = event.get('timestamp')
                elif etype == 'player_leave':
                    entry['leave_time'] = event.get('timestamp')

        # compute distances
        for pname, pdata in player_analysis_data.items():
            positions = pdata['positions']
            if len(positions) > 1:
                total_distance = 0
                for i in range(1, len(positions)):
                    prev = positions[i-1]
                    curr = positions[i]
                    try:
                        distance = ((curr['x'] - prev['x'])**2 + (curr['y'] - prev['y'])**2 + (curr['z'] - prev['z'])**2) ** 0.5
                    except Exception:
                        distance = 0
                    total_distance += distance
                pdata['total_distance'] = round(total_distance, 2)

        # Generate analysis per player using AI client
        analyses = {}
        for pname, pdata in player_analysis_data.items():
            summary = f"Player Activity Summary:\n- Total positions recorded: {len(pdata['positions'])}\n- Total distance traveled: {pdata.get('total_distance', 0)} blocks\n- Blocks placed: {len(pdata['blocks_placed'])}\n- Blocks broken: {len(pdata['blocks_broken'])}\n- Session duration: {pdata['join_time']} to {pdata['leave_time'] or 'still active'}\n\nBlock Placement Details:\n"
            for block in pdata['blocks_placed'][:10]:
                pos = block.get('position') or {'x':0,'y':0,'z':0}
                summary += f"- {block.get('type')} at ({pos.get('x')}, {pos.get('y')}, {pos.get('z')})\n"
            if len(pdata['blocks_placed']) > 10:
                summary += f"... and {len(pdata['blocks_placed']) - 10} more blocks\n"
            summary += "\nBlock Breaking Details:\n"
            for block in pdata['blocks_broken'][:10]:
                pos = block.get('position') or {'x':0,'y':0,'z':0}
                summary += f"- {block.get('type')} at ({pos.get('x')}, {pos.get('y')}, {pos.get('z')})\n"
            if len(pdata['blocks_broken']) > 10:
                summary += f"... and {len(pdata['blocks_broken']) - 10} more blocks\n"

            prompt = f"""
Please analyze the following Minecraft player's gameplay data against the provided rubric.

RUBRIC:
{rubric_content}

PLAYER: {pname}

{summary}

Please provide a detailed assessment of this player's performance based on the rubric criteria. Include specific examples from their gameplay data and suggestions for improvement. Format the response in a clear, structured way with sections for different rubric criteria.
"""

            try:
                analysis_text = await analyze_prompt(prompt)
            except Exception as e:
                analysis_text = f"Error analyzing player: {e}"

            analyses[pname] = analysis_text

        state.latest_assessment_results = {'analyses': analyses}
        return {'analyses': analyses}

    except Exception as e:
        logger.error(f"Error in analyze_player_data: {e}")
        return {'error': str(e)}

File: electron/server/session.ts
import fs from 'fs';
import path from 'path';
import state from './state';
import * as aiClient from './ai_client';

// Ensure data directory exists
const DATA_DIR = process.env.DATA_DIR || path.resolve(__dirname, '..', 'data');
if (!fs.existsSync(DATA_DIR)) fs.mkdirSync(DATA_DIR, { recursive: true });

function nowUtcIso(): string {
  return new Date().toISOString();
}

export function startSession(user: string = 'juedwards'): void {
  state.sessionStartTime = new Date();
  const iso = nowUtcIso();
  state.sessionId = `minecraft_session_${iso.replace(/[:.]/g, '_')}`;
  state.sessionEvents = [];
  const initialEvent = { timestamp: iso, event_type: 'session_start', data: { session_id: state.sessionId, server_version: '1.0', user } };
  state.sessionEvents.push(initialEvent);
  state.sessionFile = path.join(DATA_DIR, `${state.sessionId}.json`);
  saveSessionRealtime();
  console.info(`Started new session: ${state.sessionId}`);
}

export function saveSessionRealtime(): void {
  if (!state.sessionFile) return;
  const iso = nowUtcIso();
  const sessionData = {
    session_info: {
      id: state.sessionId,
      start_time: state.sessionStartTime ? (state.sessionStartTime as any).toISOString ? (state.sessionStartTime as any).toISOString() : String(state.sessionStartTime) : null,
      last_update: iso,
      duration_seconds: state.sessionStartTime ? ((Date.now() - new Date(state.sessionStartTime).getTime()) / 1000) : 0,
      total_events: state.sessionEvents ? state.sessionEvents.length : 0,
      user: 'juedwards',
      status: (state.activePlayers && state.activePlayers.size > 0) ? 'active' : 'ended',
    },
    events: state.sessionEvents || [],
  };
  const tmp = `${state.sessionFile}.tmp`;
  fs.writeFileSync(tmp, JSON.stringify(sessionData, null, 2), { encoding: 'utf8' });
  try { if (fs.existsSync(state.sessionFile)) fs.unlinkSync(state.sessionFile); } catch (e) {}
  fs.renameSync(tmp, state.sessionFile);
  state.lastSaveTime = Date.now() / 1000;
}

export function endSession(): void {
  if (!state.sessionId) return;
  const now = nowUtcIso();
  state.sessionEvents.push({ timestamp: now, event_type: 'session_end', data: { session_id: state.sessionId, duration_seconds: state.sessionStartTime ? ((Date.now() - new Date(state.sessionStartTime).getTime()) / 1000) : 0, total_events: state.sessionEvents.length } });
  saveSessionRealtime();
  console.info(`Session ended and saved: ${state.sessionFile} (${state.sessionEvents.length} events)`);
}

export function recordEvent(eventType: string, data: any): void {
  // Auto-start a session if none exists so events aren't silently dropped
  if (!state.sessionId) {
    try {
      startSession();
      console.info('Auto-started session because recordEvent was called without an active session');
    } catch (e) {
      console.error('Failed to auto-start session in recordEvent', e);
    }
  }
  const ev = { timestamp: nowUtcIso(), event_type: eventType, data };
  state.sessionEvents.push(ev);
  state.eventBuffer.push(ev);
  const now = Date.now() / 1000;
  if ((now - (state.lastSaveTime || 0) > 5) || (state.eventBuffer.length >= 50)) {
    saveSessionRealtime();
    state.eventBuffer = [];
    state.lastSaveTime = now;
  }
}

export async function analyzePlayerData(): Promise<any> {
  try {
    // Locate rubric.md in a few candidate locations (electron dev vs packaged)
    const candidatePaths = [
      path.resolve(__dirname, '..', 'rubric.md'),           // electron/rubric.md -> repo/rubric.md
      path.resolve(process.cwd(), 'rubric.md'),              // cwd/rubric.md
      path.resolve(__dirname, '..', '..', 'rubric.md'),       // electron/../.. fallback
    ];

    let rubricPath: string | null = null;
    for (const p of candidatePaths) {
      if (fs.existsSync(p)) { rubricPath = p; break; }
    }

    let rubricContent: string | null = null;
    if (rubricPath) {
      rubricContent = fs.readFileSync(rubricPath, { encoding: 'utf8' });
      console.info('Using rubric from', rubricPath);
    } else {
      // Fallback default rubric to avoid hard errors in environments where the
      // external file isn't present (development, packaged asar, etc.). You
      // should include a rubric.md in the repository or package for production.
      console.warn('Rubric file not found in candidate paths; using default embedded rubric');
      rubricContent = `Default Rubric:\n- Collaboration\n- Problem Solving\n- Use of Tools\n- Creativity\n\n(Replace this file at project root with a full rubric.md)`;
    }

    // Aggregate events similarly to Python implementation
    const playerAnalysis: Record<string, any> = {};
    console.info(`analyzePlayerData: sessionEvents=${(state.sessionEvents||[]).length}, activePlayers=${(state.activePlayers ? state.activePlayers.size : 0)}, sessionId=${state.sessionId}`);
    for (const ev of state.sessionEvents || []) {
      const etype = ev.event_type;
      const data = ev.data || {};
      if (['player_position','block_placed','block_broken','player_join','player_leave','player_chat'].includes(etype)) {
        const pid = data.player_id || data.player_name || 'unknown';
        if (!playerAnalysis[pid]) playerAnalysis[pid] = { positions: [], blocks_placed: [], blocks_broken: [], join_time: null, leave_time: null, total_distance: 0 };
        const rec = playerAnalysis[pid];
        if (etype === 'player_position') { if (data.position) rec.positions.push(data.position); }
        else if (etype === 'block_placed') rec.blocks_placed.push({ type: data.block_type, position: data.estimated_block_position, time: ev.timestamp });
        else if (etype === 'block_broken') rec.blocks_broken.push({ type: data.block_type, position: data.estimated_block_position, time: ev.timestamp });
        else if (etype === 'player_join') rec.join_time = ev.timestamp;
        else if (etype === 'player_leave') rec.leave_time = ev.timestamp;
      }
    }

    // If no player data was found, return a helpful error so the UI can explain why nothing is being analyzed
    if (Object.keys(playerAnalysis).length === 0) {
      const errMsg = `No player events recorded for analysis. sessionEvents=${(state.sessionEvents||[]).length}, activePlayers=${(state.activePlayers ? state.activePlayers.size : 0)}, playerPositions=${Object.keys(state.playerPositions||{}).length}. Ensure players are connected and generating events, and that a session has been started.`;
      console.warn(errMsg);
      return { analyses: {}, error: errMsg };
    }

    // compute distances
    for (const pid of Object.keys(playerAnalysis)) {
      const p = playerAnalysis[pid];
      if (p.positions.length > 1) {
        let total = 0;
        for (let i = 1; i < p.positions.length; i++) {
          const a = p.positions[i-1], b = p.positions[i];
          if (!a || !b) continue;
          const dx = (b.x - a.x), dy = (b.y - a.y), dz = (b.z - a.z);
          total += Math.sqrt(dx*dx + dy*dy + dz*dz);
        }
        p.total_distance = Math.round(total * 100) / 100;
      }
    }

    const analyses: Record<string, string> = {};
    for (const pid of Object.keys(playerAnalysis)) {
      const p = playerAnalysis[pid];
      let summary = `Player Activity Summary:\n- Total positions recorded: ${p.positions.length}\n- Total distance traveled: ${p.total_distance || 0} blocks\n- Blocks placed: ${p.blocks_placed.length}\n- Blocks broken: ${p.blocks_broken.length}\n- Session duration: ${p.join_time || 'N/A'} to ${p.leave_time || 'still active'}\n\nBlock Placement Details:\n`;
      for (let i = 0; i < Math.min(10, p.blocks_placed.length); i++) { const b = p.blocks_placed[i]; const pos = b.position || {x:0,y:0,z:0}; summary += `- ${b.type} at (${pos.x}, ${pos.y}, ${pos.z})\n`; }
      if (p.blocks_placed.length > 10) summary += `... and ${p.blocks_placed.length - 10} more blocks\n`;
      summary += '\nBlock Breaking Details:\n';
      for (let i = 0; i < Math.min(10, p.blocks_broken.length); i++) { const b = p.blocks_broken[i]; const pos = b.position || {x:0,y:0,z:0}; summary += `- ${b.type} at (${pos.x}, ${pos.y}, ${pos.z})\n`; }
      if (p.blocks_broken.length > 10) summary += `... and ${p.blocks_broken.length - 10} more blocks\n`;

      const prompt = `Please analyze the following Minecraft player's gameplay data against the provided rubric.\n\nRUBRIC:\n${rubricContent}\n\nPLAYER: ${pid}\n\n${summary}\n\nPlease provide a detailed assessment of this player's performance based on the rubric criteria. Include specific examples from their gameplay data and suggestions for improvement. Format the response in a clear, structured way with sections for different rubric criteria.`;

      try {
        const text = await aiClient.analyzePrompt(prompt, process.env.AZURE_OPENAI_DEPLOYMENT_NAME);
        analyses[pid] = text;
      } catch (err) {
        console.error('AI analyze failed for player', pid, err);
        analyses[pid] = `Error: ${err}`;
      }
    }

    state.latest_assessment_results = { analyses };
    return { analyses };
  } catch (err) {
    console.error('analyzePlayerData error', err);
    return { error: String(err) };
  }
}
